enhancements/installer/aws-lb-subnet-selection.md

```diff
@@ -0,0 +1,1078 @@
+// +kubebuilder:validation:MaxLength=24
+// +kubebuilder:validation:Pattern=`^subnet-[0-9A-Za-z]+$`
+type AWSSubnetID string
+
+// SubnetRole specifies the roles (aka functions) that the subnet will provide in the cluster.
+type SubnetRole string
+
+const (
+    // ClusterNodeSubnetRole specifies subnets that will be used as subnets for the
+    // control plane and compute nodes.
+    ClusterNodeSubnetRole SubnetRole = "ClusterNode"
+
+    // IngressControllerLBSubnetRole specifies subnets used by the default IngressController.
+    IngressControllerLBSubnetRole SubnetRole = "IngressControllerLB"
+
+    // ControlPlaneExternalLBSubnetRole specifies subnets used by the external control plane
+    // load balancer that serves the Kubernetes API server.
+    ControlPlaneExternalLBSubnetRole SubnetRole = "ControlPlaneExternalLB" 
+
+    // ControlPlaneInternalLBSubnetRole specifies subnets used by the internal control plane
+    // load balancer that serves the Kubernetes API server.
+    ControlPlaneInternalLBSubnetRole SubnetRole = "ControlPlaneInternalLB"
+)
+```
+
+The `platform.aws.vpc.subnets` field will be introduced under the TechPreview feature gate, and `platform.aws.subnets`
+will be deprecated when `platform.aws.vpc.subnets` graduates to GA.
+
+##### Installer Validation Rules
+
+These validation rules only apply to the new `platform.aws.vpc.subnets` field. These validations aim to improve
+user experience, but are not permanent and may be updated or removed to accommodate future use cases.
+
+###### All or Nothing Subnet Roles Selection
+
+This proposed API follows an "all or nothing" approach for roles: either all roles must be explicitly assigned
+(manual role selection), or no roles must be assigned (automatic role selection). The installer must not allow a mix of
+automatic and manual role assignments, as this could lead to confusing behavior. See [Allowing Mixed Automatic and Manual Subnet Selection Alternative](#allowing-mixed-automatic-and-manual-subnet-selection-alternative) for the reason behind this validation.
+
+However, one exception is if the cluster is private (`installconfig.publish: Internal`), then
+ControlPlaneExternalLB can be omitted.
+
+###### All Subnets Belong to the Same VPC Validation
+
+The [existing validation](https://github.com/openshift/installer/blob/0d77aa8df5ddc68e926aa11da24a87981021b256/pkg/asset/installconfig/aws/subnet.go#L91)
+for the deprecated `platform.aws.subnets` field ensuring all subnets belong to the same VPC
+should apply to all subnets specified in `platform.aws.vpc.subnets`.
+
+###### Multiple Load Balancer Subnets in the Same AZ Validation
+
+The installer must reject multiple IngressControllerLB subnets in the same AZ as this will be
+rejected by the AWS CCM.
+
+###### The Old and New Subnets Fields Cannot be Specified Together Validation
+
+Since `platform.aws.subnets` will be deprecated, the installer must not allow `platform.aws.vpc.subnets` and
```

Comment:
When both are specified, configuration provided in `platform.aws.subnets` would be ignored and `platform.aws.vpc.subnets` would be used?

Reply:
The EP currently says both specified together will be rejected. If you think it should continue regardless, let me know. To me, it felt safest to just consider it an invalid configuration. See https://github.com/openshift/enhancements/blob/8f4042b627eac2287c28e2eb871fee55c5b9cc93/enhancements/installer/aws-lb-subnet-selection.md?plain=1#L435-L438

Reply:
@sadasu are you referring to [`ConvertInstallConfig`](https://github.com/openshift/installer/blob/85a3545ebcb9166a6fa8986747dfd6e878860bb3/pkg/types/conversion/installconfig.go#L26) and the "upconvert" process? I'm just seeing that now.

I suppose converting `platform.aws.subnets` to the new `platform.aws.vpc.subnets` is pretty easy. Is the motivation for upconverting to keep the installer code simpler? So it doesn't have to maintain two separate code paths for each API?

Reply:
I think upconversion and the validation as described (reject if both are supplied) is the right way of handling it. 

To deprecate: we replace all usage of `subnets` with `vpc.subnets`, upconvert `subnets` anytime it is supplied, then after a few releases remove `subnets`

Reply:
Got it. I'll add a small snippet about the upconversion process, though I know I'm getting a bit deep in implementation details at this point.

> then after a few releases remove subnets

Hmm I have the non-goal I might need to revise: https://github.com/openshift/enhancements/blob/8f4042b627eac2287c28e2eb871fee55c5b9cc93/enhancements/installer/aws-lb-subnet-selection.md#L245

I was under the impression we must bump API versions to remove a field, since it's a breaking change. I think the InstallConfig has been on `v1` for a long time now. Does the InstallConfig use different rules on API compatibility here?

Reply:
For past deprecations the installconfig has ignored the API rules for version bumps when removing a field. 

I don’t think we need to remove the field in this case, so we could tweak my deprecation outline to something like

> then remove subnets at the next installconfig version bump